xtra-libp2p/src/endpoint.rs

```diff
@@ -155,15 +157,44 @@ pub enum Error {
     #[error("Timeout in protocol negotiation")]
     NegotiationTimeoutReached,
     #[error("Failed to negotiate protocol")]
-    NegotiationFailed(#[from] NegotiationError), // TODO(public-api): Consider breaking this up.
+    NegotiationFailed(#[from] NegotiationError),
+    // TODO(public-api): Consider breaking this up.
     #[error("Bad connection")]
-    BadConnection(#[from] yamux::ConnectionError), // TODO(public-api): Consider removing this.
+    BadConnection(#[from] yamux::ConnectionError),
+    // TODO(public-api): Consider removing this.
     #[error("Address {0} does not end with a peer ID")]
     NoPeerIdInAddress(Multiaddr),
     #[error("Either currently connecting or already connected to peer {0}")]
     AlreadyConnected(PeerId),
 }
 
+/// Subscribers that get notified on connection changes
+///
+/// This allows other actors to get notified on connection changes such as a new connection being established or dropped as well as listening addresses being added or removed.
+#[derive(Default)]
+pub struct Subscribers {
+    connection_established: Vec<Box<dyn MessageChannel<ConnectionEstablished>>>,
+    connection_dropped: Vec<Box<dyn MessageChannel<ConnectionDropped>>>,
+    listen_address_added: Vec<Box<dyn MessageChannel<ListenAddressAdded>>>,
+    listen_address_removed: Vec<Box<dyn MessageChannel<ListenAddressRemoved>>>,
```

Comment:
can you elaborate? these message channels are orthogonal, different entities can subscribe there.

also: `fan_out::Actor` does not exist in the codebase anymore.

Reply:
On second thought, I'm not sure it would add value in this case. We would achieve exactly the same functionality.

I guess the point would be that the `Endpoint` could be oblivious of the fact that it was communicating with multiple actors. And we wouldn't have to iterate through the subscribers here. But it's very natural in this case.